File: Day266/findInMountainArray.go
package main

import (
	"fmt"
)

func lSearch (arr []int, target int, low int, high int) int {
	for low <= high {
		mid := high- ((high - low) >> 1)
		if arr[mid] == target {
			return mid
		}
		if arr[mid] < target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return -1
}

func rSearch (arr []int, target int, low int, high int) int {
	for low <= high {
		mid := high - ((high - low ))
		if arr[mid] > target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return -1
}

func findInMountainArray (arr []int, target int) int {
	low, high := 0, len(arr)-1
	for low < high {
		mid := low + (high - low) /2
		if arr[mid] < arr[mid+1] {
			low = mid +1
		} else {
			high = mid -1
		}
	}
	test := lSearch(arr, target, 0, low);
	if test == -1 {
		return rSearch(arr, target, low + 1, len(arr) -1)
	}
	return test;
}


func main() {
	arr1 := []int{1,2,3,4,5,3,1}
	target1 := 3
	arr2 := []int{0,1,2,4,2,1}
	target2 := 3
	arr3 := []int{3}
	target3 := 3
	arr4 := []int{1,2,3,5,3};
	target4 := 0
	fmt.Println(findInMountainArray(arr1, target1))
	fmt.Println(findInMountainArray(arr2, target2))
	fmt.Println(findInMountainArray(arr3, target3))
	fmt.Println(findInMountainArray(arr4, target4))

}

File: Day266/findInMountainArray.java
package Day266;

class findInMountainArray {

    public static int lSearch (int[] arr, int target,int low, int high) {
        while (low <= high) {
            int mid = high - ((high - low) >> 1);
            if (arr[mid] == target) {
                return mid;
            }
            if (arr[mid] < target) {
                low  = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }

    public static int rSearch (int[] arr, int target, int low, int high) {
        while (low <= high) {
            int mid = high - ((high - low) >> 1); 
            if (arr[mid] == target) {
                return mid;
            }
            if (arr[mid] > target) {
                low = mid + 1;
            }
             else {
                high = mid -1;
             }

        }
        return -1;
    }

    public static int find(int arr[]) {
        int l =0, h = arr.length - 1;
        while ( l<h) {
            int mid =  l + (h - l )/2;
            if (arr[mid] < arr[mid+1]) {
                l = mid + 1;
            } else {
                h = mid;
            }
        }
        return l;
    }

    public static int fImMArray(int[] arr, int target) {

        int max = find(arr);

        int test = lSearch(arr, target , 0, max);
        return test == -1? rSearch(arr, target, max+1, arr.length -1): test;
    }

    public static void main(String[] args) {
        int arr1[] = {1,2,3,4,5,3,1};
        int target1 = 3;
        int arr2[] = {0,1,2,4,2,1};
        int target2 = 3;
        int arr3[] = {3};
        int target3 = 3;
        int arr4[] = {1, 2, 3, 5, 3};
        int target4 = 0;
        System.out.println(fImMArray(arr1, target1));
        System.out.println(fImMArray(arr2, target2));
        System.out.println(fImMArray(arr3, target3));
        System.out.println(fImMArray(arr4, target4));
    }
}